### PR TITLE
[DO NOT MERGE] ROC-2311 defendant how much you believe you owe

### PR DESCRIPTION
**Pull request by rattur**
Defendant partial response how much money believed you owe

### JIRA link ###

https://tools.hmcts.net/jira/browse/ROC-2311

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/features/response/form/models/howMuchOwed.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+export class HowMuchOwed implements Serializable<HowMuchOwed> {
+
+  @IsDefined({ message: ValidationErrors.AMOUNT_REQUIRED })
+  @IsPositive({ message: ValidationErrors.VALID_OWED_AMOUNT_REQUIRED })
+  amount?: number
+  @IsDefined({ message: ValidationErrors.NOT_OWE_FULL_AMOUNT_REQUIRED })
+  @IsNotBlank({ message: ValidationErrors.NOT_OWE_FULL_AMOUNT_REQUIRED })
+  @MaxLength(99000, { message: ValidationErrors.REASON_NOT_OWE_MONEY_TOO_LONG })
+  text?: string
+  constructor (amount?: number, text?: string) {
```
Comment (timja-kainos) [inline src/main/features/response/form/models/howMuchOwed.ts]: add new line above here

File: src/main/features/response/form/models/howMuchOwed.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    this.amount = amount
+    this.text = text
+  }
+
+  static fromObject (value?: any): HowMuchOwed {
+    if (value) {
+      const amount = value.amount ? parseFloat(value.amount) : undefined
+      const text = value.text
+      return new HowMuchOwed(amount, text)
+    }else {
```
Comment (timja-kainos) [inline src/main/features/response/form/models/howMuchOwed.ts]: space here

File: src/main/features/response/routes/how-much-owed.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+async function renderView (form: Form<HowMuchOwed>, res: express.Response) {
+  const user: User = res.locals.user
+  const claim: Claim = await ClaimStoreClient.retrieveLatestClaimByDefendantId(user.id)
+  const amount: number = claim.claimData.amount.totalAmount()
+  res.render(Paths.defendantHowMuchOwed.associatedView, { form: form, amount: amount, claim: claim })
+}
+
+export default express.Router()
+  .get(Paths.defendantHowMuchOwed.uri, (req: express.Request, res: express.Response) => {
```
Comment (timja-kainos) [inline src/main/features/response/routes/how-much-owed.ts]: ErrorHandling.apply()

File: src/main/features/response/form/models/howMuchOwed.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+export class HowMuchOwed implements Serializable<HowMuchOwed> {
+
+  @IsDefined({ message: ValidationErrors.AMOUNT_REQUIRED })
+  @IsPositive({ message: ValidationErrors.VALID_OWED_AMOUNT_REQUIRED })
+  amount?: number
+  @IsDefined({ message: ValidationErrors.NOT_OWE_FULL_AMOUNT_REQUIRED })
+  @IsNotBlank({ message: ValidationErrors.NOT_OWE_FULL_AMOUNT_REQUIRED })
+  @MaxLength(99000, { message: ValidationErrors.REASON_NOT_OWE_MONEY_TOO_LONG })
```
Comment (yarpo) [inline src/main/features/response/form/models/howMuchOwed.ts]: can you please add ValidationConstatnts class to keep all consts together (you can reuse it in tests then)

File: src/main/features/response/paths.ts [hunk 5]
```diff
@@ -23,4 +23,6 @@ export class Paths {
   static readonly receiptReceiver = new RoutablePath('/response/:externalId/receipt')
   // Added in case anyone has a printed copy of a PDF with the old URL
   static readonly legacyDashboardRedirect = new RoutablePath('/response/dashboard')
+  static readonly defendantHowMuchOwed = new RoutablePath('/response/how-much-owed')
```
Comment (yarpo) [inline src/main/features/response/paths.ts]: no :externalId?

File: src/main/features/response/tasks/howMuchOwedTask.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { ResponseDraft } from 'response/draft/responseDraft'
+
+export class HowMuchOwedTask {
+  static isCompleted (responseDraft: ResponseDraft): boolean {
+    if (!responseDraft.howMuchOwed.amount && responseDraft.howMuchOwed.text) {
```
Comment (yarpo) [inline src/main/features/response/tasks/howMuchOwedTask.ts]:
```
return !(!responseDraft.howMuchOwed.amount && responseDraft.howMuchOwed.text)
```

File: src/test/features/response/form/models/howMuchOwed.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      const errors = validator.validateSync(new HowMuchOwed(300, randomstring.generate(9900)))
+      expect(errors.length).to.equal(0)
+    })
+
+    it('should accept valid how much owed text', () => {
+      const errors = validator.validateSync(new HowMuchOwed(300, 'i don’t owe full amount of £300'))
+      expect(errors.length).to.equal(0)
+    })
+  })
+})
```
Comment (yarpo) [inline src/test/features/response/form/models/howMuchOwed.ts]: can you please add unit test file for howMuchOwedTask::isCompleted?

File: src/main/features/response/form/models/howMuchOwed.ts (hunk 4 above)
Comment (rattur) [reply src/main/features/response/form/models/howMuchOwed.ts]: Is it not a good idea to reuse existing validation annotations?

Comment (yarpo) [reply src/main/features/response/form/models/howMuchOwed.ts]: I think we can extract ValidationConstants from one of the files it already exist to some common module and reuse it here. 

I'd rather not use class from other local module as it creates weird dependency.

Comment (rattur) [reply src/main/features/response/form/models/howMuchOwed.ts]: Ok fair good point

File: src/main/features/response/form/models/howMuchOwed.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { Serializable } from 'models/serializable'
+import { IsDefined, IsPositive, MaxLength } from 'class-validator'
+import { IsNotBlank } from 'app/forms/validation/validators/isBlank'
+import { ValidationErrors } from 'features/validationErrors'
+
+export class HowMuchOwed implements Serializable<HowMuchOwed> {
+
+  @IsDefined({ message: ValidationErrors.AMOUNT_REQUIRED })
+  @IsPositive({ message: ValidationErrors.VALID_OWED_AMOUNT_REQUIRED })
+  amount?: number
```
Comment (rlewan) [inline src/main/features/response/form/models/howMuchOwed.ts]: Please add a newline between fields for readability.

File: src/main/features/response/form/models/howMuchOwed.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  static fromObject (value?: any): HowMuchOwed {
+    if (value) {
+      const amount = value.amount ? parseFloat(value.amount) : undefined
+      const text = value.text
+      return new HowMuchOwed(amount, text)
+    } else {
+      return new HowMuchOwed()
+    }
+  }
```
Comment (rlewan) [inline src/main/features/response/form/models/howMuchOwed.ts]: Missing newline after the curly brace.

File: src/main/features/response/draft/responseDraft.ts [hunk 10]
```diff
@@ -29,6 +31,7 @@ export class ResponseDraft extends Draft implements Serializable<ResponseDraft>
       this.counterClaim = CounterClaim.fromObject(input.counterClaim)
       this.defendantDetails = new Defendant().deserialize(input.defendantDetails)
       this.lastUpdateTimestamp = input.lastUpdateTimestamp
+      this.howMuchOwed = HowMuchOwed.fromObject(input.howMuchOwed)
```
Comment (rlewan) [inline src/main/features/response/draft/responseDraft.ts]: The pattern we follow is to use `fromObject` for form values conversion and `deserialize` for deserializing values retrieved from API. Please update this to use `deserialize`.

File: src/main/features/response/form/models/howMuchOwed.ts (hunk 8 above)
Comment (rlewan) [reply src/main/features/response/form/models/howMuchOwed.ts]: The validation on this field needs updating, please see `ClaimAmountRow#amount` for reference.

File: src/main/features/response/tasks/howMuchOwedTask.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { ResponseDraft } from 'response/draft/responseDraft'
+
+export class HowMuchOwedTask {
+  static isCompleted (responseDraft: ResponseDraft): boolean {
+    return (responseDraft.howMuchOwed.amount && responseDraft.howMuchOwed.text.length > 0)
+  }}
```
Comment (rlewan) [inline src/main/features/response/tasks/howMuchOwedTask.ts]: Please move the second curly brace to a newline as it closes the class.

File: src/main/features/response/views/how-much-owed.njk [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+{% set heading = t('How much money do you believe you owe?') %}
+
+{% set claimantFullName = claim.claimData.claimant.name %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <form novalidate method="post">
+        {{ csrfProtection(csrf) }}
+        <p>{{ t('The amount being claimed is') }} £{{ amount }}</p>
```
Comment (rlewan) [inline src/main/features/response/views/how-much-owed.njk]: Please use the `numeral` filter to format money, see `src/main/features/claim/views/check-and-send.njk` for reference.

File: src/main/features/response/views/how-much-owed.njk [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+        {{ poundInput(label = 'How much money do you believe you owe?', name = 'amount', form = form, bold = true) }}
```
Comment (rlewan) [inline src/main/features/response/views/how-much-owed.njk]: Will the label be translated in `poundInput` macro?

File: src/test/features/response/form/models/howMuchOwed.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+    it('should return an instance from given object', () => {
+      const description: string = 'I do not owe full amount'
+      const amount: number = 300
+      const result: HowMuchOwed = new HowMuchOwed().deserialize({
+        amount: amount,
+        text: description
+      })
+      expect(result.amount).to.be.equals(amount)
+      expect(result.text).to.be.equals(description)
```
Comment (rlewan) [inline src/test/features/response/form/models/howMuchOwed.ts]: Would be nice if you could change these two to `.to.be.equal` to make it read correctly.

File: src/main/features/response/views/how-much-owed.njk (hunk 13 above)
Comment (rattur) [reply src/main/features/response/views/how-much-owed.njk]: Yes

**Pull request merged**